presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java

```diff
@@ -345,16 +447,378 @@ public void testCtasAcidTable(boolean isPartitioned, BucketingType bucketingType
+                    tableName, bucketed ? "CLUSTERED BY (col2) INTO 3 BUCKETS" : ""));
+
+            execute(inserter1, format("INSERT INTO %s VALUES (1, 100, 'a'), (2, 200, 'b')", tableName));
+            verifySelectForPrestoAndHive("SELECT * FROM " + tableName, "true", row(1, 100, "a"), row(2, 200, "b"));
+
+            execute(inserter2, format("INSERT INTO %s VALUES (3, 300, 'c'), (4, 400, 'a')", tableName));
+            verifySelectForPrestoAndHive("SELECT * FROM " + tableName, "true", row(1, 100, "a"), row(2, 200, "b"), row(3, 300, "c"), row(4, 400, "a"));
+
+            execute(inserter1, format("INSERT INTO %s VALUES (5, 500, 'b'), (6, 600, 'c')", tableName));
+            verifySelectForPrestoAndHive("SELECT * FROM " + tableName, "true", row(1, 100, "a"), row(2, 200, "b"), row(3, 300, "c"), row(4, 400, "a"), row(5, 500, "b"), row(6, 600, "c"));
+            verifySelectForPrestoAndHive("SELECT * FROM " + tableName, "col2 > 300", row(4, 400, "a"), row(5, 500, "b"), row(6, 600, "c"));
+
+            execute(inserter2, format("INSERT INTO %s VALUES (7, 700, 'b'), (8, 800, 'c')", tableName));
+            verifySelectForPrestoAndHive("SELECT * FROM " + tableName, "true", row(1, 100, "a"), row(2, 200, "b"), row(3, 300, "c"), row(4, 400, "a"), row(5, 500, "b"), row(6, 600, "c"), row(7, 700, "b"), row(8, 800, "c"));
+            verifySelectForPrestoAndHive("SELECT * FROM " + tableName, "col3 = 'c'", row(3, 300, "c"), row(6, 600, "c"), row(8, 800, "c"));
+        });
+    }
+
+    @DataProvider
+    public Object[][] insertersProvider()
+    {
+        return new Object[][] {
+                {false, HiveOrPresto.HIVE, HiveOrPresto.PRESTO},
+                {false, HiveOrPresto.PRESTO, HiveOrPresto.PRESTO},
+                {true, HiveOrPresto.HIVE, HiveOrPresto.PRESTO},
+                {true, HiveOrPresto.PRESTO, HiveOrPresto.PRESTO},
+        };
+    }
+
+    private enum HiveOrPresto
+    {
+        HIVE,
+        PRESTO
+    }
+
+    private static QueryResult execute(HiveOrPresto hiveOrPresto, String sql, QueryExecutor.QueryParam... params)
+    {
+        return executorFor(hiveOrPresto).executeQuery(sql, params);
+    }
+
+    private static QueryExecutor executorFor(HiveOrPresto hiveOrPresto)
+    {
+        switch (hiveOrPresto) {
+            case HIVE:
+                return onHive();
+            case PRESTO:
+                return onPresto();
+            default:
+                throw new IllegalStateException("Unknown enum value " + hiveOrPresto);
+        }
+    }
+
+    @DataProvider
+    public Object[][] inserterAndDeleterProvider()
+    {
+        return new Object[][] {
+                {HiveOrPresto.HIVE, HiveOrPresto.PRESTO},
+                {HiveOrPresto.PRESTO, HiveOrPresto.PRESTO},
+                {HiveOrPresto.PRESTO, HiveOrPresto.HIVE}
+        };
+    }
+
+    void withTemporaryTable(String rootName, boolean transactional, boolean isPartitioned, BucketingType bucketingType, Consumer<String> testRunner)
```

Comment:
I like this help function, and the order and economy it brings to the tests I've added.  If there is no objection, I'll follow this PR with a PR that makes no semantic change but updates the previously existing tests to use this helper function.

Reply:
> Consumer<String> testRunner

instead of putting the test in a lambda (which has some drawbacks), you can use t-w-r like we do in other tests --
https://github.com/prestosql/presto/blob/fbd4ca7f6ca1326a077fd06cd33b687cbfe17cf2/presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java#L414

Reply:
I see your point, @findepi.  However, what I like about the lambda formulation is that it doesn't dictate the form of the `CREATE TABLE` statement; it just supplies the temporary table name.  In the tests I wrote, the `CREATE TABLE` statements vary over a pretty wide range, and don't always fit into the formulation (bucketed | unbucketed) X (partitioned | unpartioned)